src/resources/spi_flash_esp32.cc

```diff
@@ -0,0 +1,285 @@
+  spi_nand_flash_device_t *_nand_flash_device = null;
+  spi_device_handle_t _nand_spi_device = null;
+#endif
+};
+
+MODULE_IMPLEMENTATION(spi_flash, MODULE_SPI_FLASH);
+
+PRIMITIVE(init_sdcard) {
+  ARGS(cstring, mount_point, SPIResourceGroup, spi_host, int, gpio_cs, int, format_if_mount_failed, int, max_files, int, allocation_unit_size)
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (!proxy) ALLOCATION_FAILED;
+
+  char* mp = static_cast<char *>(malloc(strlen(mount_point)));
+  if (!mp) MALLOC_FAILED;
+  strcpy(mp, mount_point);
+
+  auto group = _new SPIFlashResourceGroup(process,  mp);
+
+  if (!group) {
+    free(mp);
+    ALLOCATION_FAILED;
+  }
+
+  sdmmc_host_t host = SDSPI_HOST_DEFAULT();
+  host.slot = spi_host->host_device();
+
+  sdspi_device_config_t slot_config = SDSPI_DEVICE_CONFIG_DEFAULT();
+  slot_config.host_id = static_cast<spi_host_device_t>(host.slot);
+  slot_config.gpio_cs = static_cast<gpio_num_t>(gpio_cs);
+
+  esp_vfs_fat_sdmmc_mount_config_t mount_config = {
+      .format_if_mount_failed = static_cast<bool>(format_if_mount_failed),
+      .max_files = max_files,
+      .allocation_unit_size = static_cast<size_t>(allocation_unit_size)
+  };
+  sdmmc_card_t* card;
+  esp_err_t ret = esp_vfs_fat_sdspi_mount(mount_point, &host, &slot_config, &mount_config, &card);
+  if (ret != ESP_OK) {
+    return Primitive::os_error(ret, process);
+  }
+
+  group->set_card(card);
+
+  proxy->set_external_address(group);
+  return proxy;
+}
+
+PRIMITIVE(init_nor_flash) {
+  ARGS(cstring, mount_point, SPIResourceGroup, spi_bus, int, gpio_cs,int, frequency, int, format_if_mount_failed, int, max_files, int, allocation_unit_size)
+
+  if (frequency < 0 || frequency > ESP_FLASH_80MHZ) INVALID_ARGUMENT;
+
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (!proxy) ALLOCATION_FAILED;
+
+  char* mp = static_cast<char *>(malloc(strlen(mount_point)));
+  if (!mp) MALLOC_FAILED;
+  strcpy(mp, mount_point);
+
+  auto group = _new SPIFlashResourceGroup(process,  mp);
+
+  if (!group) {
+    free(mp);
+    ALLOCATION_FAILED;
+  }
+
+  esp_flash_spi_device_config_t conf = {
+      .host_id = spi_bus->host_device(),
+      .cs_io_num = gpio_cs,
+      .io_mode = SPI_FLASH_FASTRD,
+      .speed = static_cast<esp_flash_speed_t>(frequency),
+      .input_delay_ns = 0,
+      .cs_id = 0
+  };
+
+  esp_flash_t *chip;
+  esp_err_t ret = spi_bus_add_flash_device(&chip, &conf);
+  if (ret != ESP_OK) {
+    return Primitive::os_error(ret, process);
+  }
+
+  group->set_chip(chip);
+
+  ret = esp_flash_init(chip);
+
+  if (ret != ESP_OK) {
+    group->close();
+    return Primitive::os_error(ret, process);
+  }
+
+  size_t size;
+  ret = esp_flash_get_size(chip, &size);
+  if (ret != ESP_OK) {
+    group->close();
+    return Primitive::os_error(ret, process);
+  }
+
+  // We are using mount_point as the label for the external partition since that should be unique when multiple NOR
+  // flash chips are used
```

Comment:
```suggestion
  // We are using mount_point as the label for the external partition since that should 
  // be unique when multiple NOR flash chips are used.
```